### PR TITLE
Add "from-empty-schema" option for "diff" command

### PR DESCRIPTION
**Pull request by guilliamxavier**
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues | #820

#### Summary

Allow to execute `migrations:diff --from-empty-schema`
(as alternative to `migrations:dump-schema` but using the mapping instead of the database)

### CODE REVIEW

**Comment by guilliamxavier** [comment]
Not tested yet (hence draft) but WDYT?

**Comment by guilliamxavier** [comment]
Update: I have tested the patch locally and I confirm it works as expected 😃 (but writing an automated test would be another story...)

File: tests/Doctrine/Migrations/Tests/Tools/Console/Command/DiffCommandTest.php [hunk 1]
```diff
@@ -55,9 +55,14 @@ public function testExecute() : void
         $input->expects(self::at(4))
             ->method('getOption')
             ->with('check-database-platform')
-            ->willReturn(true);
+            ->willReturn(false);
```
Comment (guilliamxavier) [inline tests/Doctrine/Migrations/Tests/Tools/Console/Command/DiffCommandTest.php:58]: This change is just to have a value different from the default (like all other tested options) and for the call below

**Comment by goetas** [comment]
I like this change! 

A test should test the `createEmptySchema` method asserting something on the expected output.
(the current changes to the tests are just making sure that args are read as expected)

**Comment by guilliamxavier** [comment]
I have added tests (but I'm not sure because I couldn't mock the `new Schema(...)` in `DiffGenerator::createEmptySchema()`...)

File: tests/Doctrine/Migrations/Tests/Generator/DiffGeneratorTest.php [hunk 2]
```diff
@@ -118,6 +119,60 @@ static function ($name) {
...
+            ->willReturn(null);
+
+        $toSchema->method('getName')->willReturn('schema');
+        $toSchema->method('getTables')->willReturn(['schema.table_name' => $table]);
+        $toSchema->method('getTable')->with('table_name')->willReturn($table);
+        $toSchema->method('getNamespaces')->willReturn([]);
+        $toSchema->method('getSequences')->willReturn([]);
+
+        $table->method('getName')->willReturn('schema.table_name');
+        $table->method('getShortestName')->with('schema')->willReturn('table_name');
```
Comment (guilliamxavier) [inline tests/Doctrine/Migrations/Tests/Generator/DiffGeneratorTest.php:152]: After thinking, that part of the test is too fragile (tightly coupled to internal DBAL implementation). I'll rework the implementation (during the weekend)

**Comment by guilliamxavier** [comment]
I don't know why the "**continuous-integration/travis-ci**" check is blocked at "_Expected — Waiting for status to be reported_"... you can see that it passed: https://travis-ci.org/github/doctrine/migrations/builds/700372177

Anyway, this is now ready for review @goetas :slightly_smiling_face:

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/2.3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

**Review by greg0ire — COMMENTED** [review]
Can you please add some documentation?

**Comment by guilliamxavier** [comment]
@greg0ire: done and done ;)

**Review by greg0ire — APPROVED** [review]

**Comment by goetas** [comment]
thanks @guilliamxavier 

I think that I will have fun times when merging back this to `master` :)

**Comment by alzee** [comment]
Thanks @guilliamxavier
Great job!

**Pull request merged**